### PR TITLE
Fix Earley parser stack overflow issue

### PR DESCRIPTION
**Pull request by mikolasstuchlik**
This pull request solves partially the https://github.com/palle-k/Covfefe/issues/11 issue. I can not close the issue outright, becaue CYK parser has still stack overflow issue present.

The new algorythm might be slower than the original recursive approach, but is not constrained by the length of the parsed sentence.

Note: tests now take few minutes to complete.

### CODE REVIEW

**Comment by mikolasstuchlik** [comment]
Turns out, the parser stack-overflows in another function.
Snímek obrazovky 2021-11-04 v 14 02 07

The test `testErleyStackOverflow` was modified to use 10000 long word and failed after 46 minutes 15 seconds of runtime.

**Comment by mikolasstuchlik** [comment]
I have been able to solve the `explode` recursion. I have also profiled the program and found out, that the major bottleneck was in linear iteration through a Set instance. I have changed the Set to Array, reducing the run time for 10000 items test from 46 minutes down to < 4 minutes.

I would like to request a review. I have decided not to add an inline documentation. Let me know what pars of the code you would like to see documented. All the machines have the same architecture.

**Comment by mikolasstuchlik** [comment]
Unfortunately, I have observed another issue related to recursion: all other recursive functions :/

When working on Linux, the `syntaxTree.description.count` was crashing due to `map` function stack overflow. I'm not going to solve the issue yet from two reasons:
 - I don't need the `syntaxTree.description` since I wont be able to generate such a magnificently big image (`dot` kept crashing on me even for 1000 letters long word anyway).
 - In case I would absolutely need the `syntaxTree.description` output, the program completed when usign `-c release` configuration.

Since I am now able to prototype in a reasonable time, I don't need to address the issue right now. If the PR is accepted, maybe I'll address it in the future.

File: Sources/Covfefe/SyntaxTree.swift [hunk 1]
```diff
@@ -34,47 +34,222 @@ public enum SyntaxTree<Element, LeafElement> {
...
+				try resolve(key: key, children: children, iteratedIndex: index)
+			}
+		}
+	}
+
+}
+
+
+public extension SyntaxTree {
+	private enum ExplostionStackFrame {
```
Comment (palle-k) [inline Sources/Covfefe/SyntaxTree.swift]: Typo in Explosion

File: Sources/Covfefe/SyntaxTree.swift [hunk 2]
```diff
@@ -83,54 +258,76 @@ public extension SyntaxTree {
...
+			let newChildren = trees.popLast()!
+			trees[trees.count - 1].append(.node(key: key, children: newChildren))
 		}
+
+		return trees.first!.first
 	}
-	
-	/// Explodes nodes and passes all child nodes to the parent node if the given closure returns true
+
+	/// Filters the tree by removing all leafs if the given predicate is false
```
Comment (palle-k) [inline Sources/Covfefe/SyntaxTree.swift]: Please use leaves as the plural of leaf.

File: Tests/CovfefeTests/PerformanceTests.swift [hunk 3]
```diff
@@ -91,7 +91,31 @@ class PerformanceTests: XCTestCase {
 		}
 	}
 
-    static var allTests = [
-        ("testEarleyPerformance", testEarleyPerformance)
-    ]
+	func testErleyStackOverflow() {
```
Comment (palle-k) [inline Tests/CovfefeTests/PerformanceTests.swift]: Fix typo in Earley

**Review by palle-k — CHANGES_REQUESTED** [review]

**Comment by mikolasstuchlik** [comment]
@palle-k Hi, can we merge this? :)

**Review by palle-k — APPROVED** [review]

**Pull request merged**